### PR TITLE
Added support for box86 on arm64 platforms

### PR DESCRIPTION
**Pull request by Jai-JAP**
Added support for box86 on arm64 platforms (As per box86 updated CMakeLists.txt)  -DRPI4ARM64 flag is not included in Compile.md Instructions

### CODE REVIEW

**Comment by Itai-Nelken** [comment]
you should use pi-apps `pkg-install` script to install the dependencies.

**Comment by Jai-JAP** [comment]
> you should use pi-apps `pkg-install` script to install the dependencies.
> and add the `purge-installed` line to the uninstall script if the arch is arm64.

**Comment by Itai-Nelken** [comment]
why did you close this?

**Comment by Jai-JAP** [comment]
> you should use pi-apps `pkg-install` script to install the dependencies.
> and add the `purge-installed` line to the uninstall script if the arch is arm64.

Present uninstall script works fine for arm64  as well

**Comment by Itai-Nelken** [comment]
>Present uninstall script works fine for arm64 as well

@Jai-JAP yes to uninstall box86 but not to uninstall the dependencies you installed to compile box86.

**Comment by Jai-JAP** [comment]
> @Jai-JAP why did you close this?

Sorry it happened accidentally

File: apps/Box86/install-64 [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+elif [ "${CODENAME}" = "groovy" ];then
+  "${DIRECTORY}/pkg-install" "libc6:armhf  libx11-6:armhf  libgdk-pixbuf2.0-0:armhf libgtk2.0-0:armhf libstdc++6:armhf libsdl2-2.0-0:armhf mesa-va-drivers:armhf libsdl1.2-dev:armhf libsdl-mixer1.2:armhf libpng16-16:armhf libcal3d12v5:armhf libsdl2-net-2.0-0:armhf libopenal1:armhf libsdl2-image-2.0-0:armhf libvorbis-dev:armhf libcurl4:armhf osspd:armhf pulseaudio:armhf libjpeg62:armhf libudev1:armhf libgl1-mesa-dev:armhf libsnappy1v5:armhf libx11-dev:armhf libsmpeg0:armhf libboost-filesystem1.71.0:armhf libboost-program-options1.71.0:armhf libavcodec58:armhf libavformat58:armhf libswscale5:armhf libmyguiengine3debian1v5:armhf libboost-iostreams1.71.0:armhf  libsdl2-mixer-2.0-0:armhf" || exit 1
+
+"${DIRECTORY}/pkg-install" "gcc-arm-linux-gnueabihf" || exit 1
```
Comment (Itai-Nelken) [inline apps/Box86/install-64:19]: add this package (`gcc-arm-linux-gbueabihf`) to the 2 lines above, I don't think you can run `pkg-install` multiple times one after the other.

File: apps/Box86/install-64 [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+if [ -d ~/box86 ];then
+  echo -e "\e[33mbox86 already exists on your system. Moving the old version to Trash...\e[39m"
+  gio trash ~/box86
+fi
+
+"${DIRECTORY}/pkg-install" "cmake" "$(dirname "$0")" || exit 1
```
Comment (Itai-Nelken) [inline apps/Box86/install-64:26]: same as my above comment.

File: apps/Box86/install-64 [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+git clone https://github.com/ptitSeb/box86 || error "failed to download repo!"
+cd box86 || error "failed to enter directory!"
+mkdir build; cd build
+cmake .. -DRPI4ARM64=1 -DCMAKE_BUILD_TYPE=RelWithDebInfo || error "cmake failed!"
+make -j4 || error "make failed!"
```
Comment (Itai-Nelken) [inline apps/Box86/install-64:32]: add this instead:
```bash
cores="$(nproc)"
make -j$cores
```
to stop pi3's from failing because they run out of memory.

**Review by Itai-Nelken — CHANGES_REQUESTED** [review]
added a few suggestions

**Comment by Itai-Nelken** [comment]
also add this code to the uninstall script:
```bash
if [ ! -z "$(file "$(readlink -f "/sbin/init")" | grep 64)" ];then
    "${DIRECTORY}/purge-installed" "$(dirname "$0")" || exit 1
elif [ ! -z "$(file "$(readlink -f "/sbin/init")" | grep 32)" ];then
    sleep 0.00001
else
    error "Can't detect OS architecture! something is very wrong!"
fi
```
it will uninstall all the packages `pkg-install` installed, but only if you are on a 64bit OS.

File: apps/Box86/uninstall [hunk 4]
```diff
@@ -12,6 +12,14 @@ cd ~/box86/build
 sudo make uninstall
 cd $HOME
 
+CODENAME=$(lsb_release -cs)
```
Comment (Itai-Nelken) [inline apps/Box86/uninstall:15]: no, thast not how `purge-installed` works (I made the same mistake once), use this instead:
```bash
if [ ! -z "$(file "$(readlink -f "/sbin/init")" | grep 64)" ]; then
    "${DIRECTORY}/purge-installed" "$(dirname "$0")" || exit 1
elif [ ! -z "$(file "$(readlink -f "/sbin/init")" | grep 32)" ]; then
    sleep 0.00001
else
    error "Can't detect OS architecture! something is very wrong!"
fi
```

File: apps/Box86/uninstall [hunk 5]
```diff
@@ -12,6 +12,14 @@ cd ~/box86/build
...
+if [ "${CODENAME}" = "focal" || "${CODENAME}" = "buster" ];then
+  "${DIRECTORY}/purge-installed" "libc6:armhf  libx11-6:armhf  libgdk-pixbuf2.0-0:armhf libgtk2.0-0:armhf libstdc++6:armhf libsdl2-2.0-0:armhf mesa-va-drivers:armhf libsdl1.2-dev:armhf libsdl-mixer1.2:armhf libpng16-16:armhf libcal3d12v5:armhf libsdl2-net-2.0-0:armhf libopenal1:armhf libsdl2-image-2.0-0:armhf libvorbis-dev:armhf libcurl4:armhf osspd:armhf pulseaudio:armhf libjpeg62:armhf libudev1:armhf libgl1-mesa-dev:armhf libsnappy1v5:armhf libx11-dev:armhf libsmpeg0:armhf libboost-filesystem1.67.0:armhf libboost-program-options1.67.0:armhf libavcodec58:armhf libavformat58:armhf libswscale5:armhf libmyguiengine3debian1v5:armhf libboost-iostreams1.67.0:armhf  libsdl2-mixer-2.0-0:armhf" || exit 1
```
Comment (Itai-Nelken) [inline apps/Box86/uninstall]: no, thast not how `purge-installed` works (I made the same mistake once), use this instead:
```bash
if [ ! -z "$(file "$(readlink -f "/sbin/init")" | grep 64)" ]; then
    "${DIRECTORY}/purge-installed" "$(dirname "$0")" || exit 1
elif [ ! -z "$(file "$(readlink -f "/sbin/init")" | grep 32)" ]; then
    sleep 0.00001
else
    error "Can't detect OS architecture! something is very wrong!"
fi
```

**Review by Itai-Nelken — CHANGES_REQUESTED** [review]
fix `purge-installed`

File: apps/Box86/uninstall (hunk 5 above)
Comment (Itai-Nelken) [inline apps/Box86/uninstall]: no, thats not how `purge-installed` works (I made the same mistake once), use this instead:
```bash
if [ ! -z "$(file "$(readlink -f "/sbin/init")" | grep 64)" ]; then
    "${DIRECTORY}/purge-installed" "$(dirname "$0")" || exit 1
elif [ ! -z "$(file "$(readlink -f "/sbin/init")" | grep 32)" ]; then
    sleep 0.00001
else
    error "Can't detect OS architecture! something is very wrong!"
fi
```

**Comment by Itai-Nelken** [comment]
@Jai-JAP you didn't understand how `purge-installed` works, you **don't** tell it what packages to remove, it knows what packages `pkg-install` installed, and uninstalls them.
again, use this in the uninstall script:
```bash
if [ ! -z "$(file "$(readlink -f "/sbin/init")" | grep 64)" ]; then
    "${DIRECTORY}/purge-installed" "$(dirname "$0")" || exit 1
elif [ ! -z "$(file "$(readlink -f "/sbin/init")" | grep 32)" ]; then
    sleep 0.00001
else
    error "Can't detect OS architecture! something is very wrong!"
fi
```

**Comment by mobilegmYT** [comment]
Correct me if I'm wrong, but box86 needs a 32-bit toolchain to work, and you need a chroot for that.

**Comment by Itai-Nelken** [comment]
> Correct me if I'm wrong, but box86 needs a 32-bit toolchain to work, and you need a chroot for that.

multiarch works as well.

**Comment by mobilegmYT** [comment]
> > Correct me if I'm wrong, but box86 needs a 32-bit toolchain to work, and you need a chroot for that.
> 
> multiarch works as well.

Oh that works? Whenever someone talked about running it on 64-bit they said to chroot in the discord.

**Comment by Itai-Nelken** [comment]
> > > Correct me if I'm wrong, but box86 needs a 32-bit toolchain to work, and you need a chroot for that.
> > 
> > 
> > multiarch works as well.
> 
> Oh that works? Whenever someone talked about running it on 64-bit they said to chroot in the discord.

works for me on Ubuntu 20.10

File: apps/BalenaEtcher/install-32 [hunk 6]
```diff
@@ -7,10 +7,10 @@ function error {
   exit 1
 }
 
-rm -f ./etcher_1.5.113_armv7l.deb
+rm -f ./etcher_1.5.114_armv7l.deb
```
Comment (Itai-Nelken) [inline apps/BalenaEtcher/install-32:10]: I suggest you do this:
```suggestion
rm -f etcher*armv7l.deb
```

**Review by Itai-Nelken — CHANGES_REQUESTED** [review]
suggestion

**Comment by mobilegmYT** [comment]
@Jai-JAP Please make a separate pull request for other apps, and also v1.1.115 isn't in itai's repo.

**Comment by Itai-Nelken** [comment]
also @Jai-JAP, this pr is for box86 on arm64, don't add etcher to it.

**Comment by Jai-JAP** [comment]
Box86 for arm64 [scripts](https://github.com/Botspot/pi-apps/tree/master/apps/Box86) (install-64 & updated uninstall scripts) by @Itai-Nelken have been added.
Closing this pull request.

**Comment by JulyIghor** [comment]
Not working on Ubuntu arm64, still tries to install armhf lib versions

**Comment by Itai-Nelken** [comment]
> Not working on Ubuntu groovy arm64, still tries to install armhf lib versions

thats what its supposed to do, also pi-apps already added box86 on 64bit in #452.


box86 only works on armhf, so to run it on arm64 multiarch is used to install the armhf dependencies and box86.

**Comment by JulyIghor** [comment]
> > Not working on Ubuntu groovy arm64, still tries to install armhf lib versions
> 
> thats what its supposed to do, also pi-apps already added box86 on 64bit in #452.
> 
> box86 only works on armhf, so to run it on arm64 multiarch is used to install the armhf dependencies and box86.

Ok, thanks for the answer

**Pull request closed without merging**